core/tee/tee_enc_fs.c

```diff
@@ -0,0 +1,761 @@
+	fdp->pos += len;
+	res = len;
+
+exit:
+	return res;
+}
+
+static tee_fs_off_t tee_fs_lseek(int fd, tee_fs_off_t offset, int whence)
+{
+	tee_fs_off_t res = -1;
+	tee_fs_off_t new_pos;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	if (!fdp)
+		goto exit;
+
+	switch (whence) {
+
+	case TEE_FS_SEEK_SET:
+		new_pos = offset;
+		break;
+
+	case TEE_FS_SEEK_CUR:
+		new_pos = fdp->pos + offset;
+		break;
+
+	case TEE_FS_SEEK_END:
+		new_pos = fdp->len + offset;
+		break;
+
+	default:
+		goto exit;
+	}
+
+	/*
+	 * file hole is not supported in this implementation,
+	 * restrict the file postion within file length
+	 * for simplicity
+	 */
+	if (new_pos > (tee_fs_off_t)fdp->len)
+		goto exit;
+
+	res = fdp->pos = new_pos;
+exit:
+	return res;
+}
+
+static int tee_fs_rename(const char *old, const char *new)
+{
+	int res = -1;
+	char *tmp = NULL;
+	struct tee_fs_rpc head = { 0 };
+	size_t len_old = strlen(old) + 1;
+	size_t len_new = strlen(new) + 1;
+	size_t len = len_old + len_new;
+
+	if (len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	tmp = malloc(len);
+	if (!tmp)
+		goto exit;
+	memcpy(tmp, old, len_old);
+	memcpy(tmp + len_old, new, len_new);
+
+	head.op = TEE_FS_RENAME;
+
+	res = tee_fs_send_cmd(&head, tmp, len, TEE_FS_MODE_IN);
+	if (res)
+		goto exit;
+
+	res = head.res;
+
+exit:
+	free(tmp);
+	return res;
+}
+
+static int tee_fs_unlink(const char *file)
+{
+	int res = -1;
+	struct tee_fs_rpc head = { 0 };
+	size_t len = strlen(file) + 1;
+
+	if (len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	head.op = TEE_FS_UNLINK;
+
+	res = tee_fs_send_cmd(&head, (void *)file, len, TEE_FS_MODE_IN);
+	if (res)
+		goto exit;
+
+	res = head.res;
+
+exit:
+	return res;
+}
+
+static int tee_fs_ftruncate(int fd, tee_fs_off_t length)
+{
+	int res = -1;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	if (fdp == NULL)
+		goto exit;
+
+	if (fdp->flags & TEE_FS_O_RDONLY)
+		goto exit;
+
+	res = update_file_size(fdp, length);
+
+exit:
+	return res;
+}
+
+static int tee_fs_mkdir(const char *path, tee_fs_mode_t mode)
+{
+	int res = -1;
+	struct tee_fs_rpc head = { 0 };
+	uint32_t len;
+
+	len = strlen(path) + 1;
+	if (len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	head.op = TEE_FS_MKDIR;
+	head.flags = mode;
+
+	res = tee_fs_send_cmd(&head, (void *)path, len, TEE_FS_MODE_IN);
+	if (res)
+		goto exit;
+
+	res = head.res;
+
+exit:
+	return res;
+}
+
+static tee_fs_dir *tee_fs_opendir(const char *name)
+{
+	struct tee_fs_rpc head = { 0 };
+	uint32_t len;
+	struct tee_fs_dir *dir = NULL;
+
+	len = strlen(name) + 1;
+	if (len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	head.op = TEE_FS_OPENDIR;
+
+	if (tee_fs_send_cmd(&head, (void *)name, len, TEE_FS_MODE_IN))
+		goto exit;
+
+	if (head.res < 0)
+		goto exit;
+
+	dir = malloc(sizeof(struct tee_fs_dir));
+	if (dir == NULL) {
+		int nw_dir = head.res;
+
+		memset(&head, 0, sizeof(head));
+		head.op = TEE_FS_CLOSEDIR;
+		head.arg = nw_dir;
+		tee_fs_send_cmd(&head, NULL, 0, TEE_FS_MODE_NONE);
```

Comment:
But many of them (`tee_file_operations`) return an integer for file descriptor. and -1 for error. We might able to define customized error code (e.g. -2, -3, -4 ....) and map them to TEE_Result. Or completely re-design the API. How to you think?